### PR TITLE
Second part of The movieDB App 

### PR DESCRIPTION
**Pull request by AlejoArea**
Added repositories folder, created repository interface to implement movie and genre repositories for Json decoding. Added movie list screen and movie details screen for navigation.

[recorded_app_module_3.webm](https://github.com/AlejoArea/TheMovieDB/assets/114313724/35fa28f4-c38f-4978-bf7f-d4e83eb0ea95)

### CODE REVIEW

File: lib/models/genre.dart [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  final String genre;
+
+  Genre({
+    required this.id,
+    required this.genre,
+  });
+
+  factory Genre.fromJson(Map<String, dynamic> json) {
+    final id = json['id'];
+    final genre = json['name'];
```
Comment (LeonardoMolleker) [inline lib/models/genre.dart]: Is not necessary to previously store the values on variables, you can send them directly to the constructor

File: lib/screens/movie_details.dart [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  static const String overviewTitle = 'Overview';
+  static const String genreTitle = "Genres";
+
+  @override
+  Widget build(BuildContext context) {
+    void incrementLikeCounter() {
+      setState(() {
+        _counter++;
+      });
+    }
```
Comment (LeonardoMolleker) [inline lib/screens/movie_details.dart]: let's move this outside the build method

File: lib/screens/movie_details.dart [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text(
+          Constants.titleApp,
+        ),
+        centerTitle: true,
+      ),
+      body: SafeArea(
+        child: ListView(
```
Comment (LeonardoMolleker) [inline lib/screens/movie_details.dart]: Why a list view if you only have one child?

File: lib/screens/movie_list.dart [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                                  ),
+                                )
+                              ],
+                            ),
+                          ),
+                        ],
+                      ),
+                    ),
+                  );
+                })
```
Comment (LeonardoMolleker) [inline lib/screens/movie_list.dart:114]: trailing comma

File: lib/widgets/genres.dart [hunk 5]
```diff
@@ -1,53 +1,89 @@
...
-              ),
+    return FutureBuilder(
+      future: allGenres = repository.getData(),
+      builder: (
+        context,
+        snapshot,
+      ) {
+        if (snapshot.hasData) {
+          List<Genre> snapshotList = snapshot.data ?? [];
+          movieGenres = repository.getGenresFromId(genresId, snapshotList);
```
Comment (LeonardoMolleker) [inline lib/widgets/genres.dart]: trailing comma

Comment (AlejoArea) [reply lib/widgets/genres.dart]: done

File: lib/screens/movie_details.dart (hunk 3 above)
Comment (AlejoArea) [reply lib/screens/movie_details.dart]: removed unnecessary listview

File: lib/screens/movie_details.dart (hunk 2 above)
Comment (AlejoArea) [reply lib/screens/movie_details.dart]: done

File: lib/models/genre.dart (hunk 1 above)
Comment (AlejoArea) [reply lib/models/genre.dart]: done

File: lib/screens/movie_details.dart (hunk 3 above)
Comment (AlejoArea) [reply lib/screens/movie_details.dart]: changed to SingleChildScrollView

File: lib/models/movie.dart [hunk 6]
```diff
@@ -1,36 +1,74 @@
...
+      originalTitle: json['original_title'],
+      originalLanguage: json['original_language'],
+      id: json['id'] as int,
+      adult: json['adult'] as bool,
+      releaseDate: json['release_date'],
+      overview: json['overview'] as String,
+      voteAverage: json['vote_average'].toDouble(),
+      genres: List<int>.from(json['genre_ids']),
+      posterPath: 'https://image.tmdb.org/t/p/w500${json['poster_path']}',
+      backdropPath: 'https://image.tmdb.org/t/p/w500${json['backdrop_path']}',
```
Comment (matiasilveiro-gl) [inline lib/models/movie.dart]: We can move the concatenation of the base url to a getter method:
`String get posterUrl => 'https://image.tmdb.org/t/p/w500${posterPath}'`

File: lib/repositories/i_repository.dart [hunk 7]
```diff
@@ -0,0 +1,3 @@
+abstract class IRepository {
+  Future<List<dynamic>> getData();
```
Comment (matiasilveiro-gl) [inline lib/repositories/i_repository.dart]: Use a generic type `<T>` for the interface. Avoid using `dynamic`

File: lib/screens/movie_details.dart [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                children: [
+                  Header(
+                    backdropPath: widget.movie.backdropPath,
+                  ),
+                  const SizedBox(
+                    height: Constants.sizedBoxHeight,
+                  ),
+                  GeneralText(
+                    generalText:
+                        '$originalTitle ${widget.movie.originalTitle.toString()}',
```
Comment (matiasilveiro-gl) [inline lib/screens/movie_details.dart]: The `.toString()` is redundant here

File: lib/widgets/genres.dart [hunk 9]
```diff
@@ -1,53 +1,89 @@
...
-              height: genrePaddingHeight,
-              width: genrePaddingWidth,
-              child: Center(
-                child: GeneralText(
-                  generalText: genres[index],
-                  fontSize: Constants.smallTextFont,
-                ),
-              ),
+    return FutureBuilder(
+      future: allGenres = repository.getData(),
```
Comment (matiasilveiro-gl) [inline lib/widgets/genres.dart]: I'd follow the approach of the `Stateful` widget you did for the movies

File: lib/widgets/genres.dart [hunk 10]
```diff
@@ -1,53 +1,89 @@
...
+      builder: (
+        context,
+        snapshot,
+      ) {
+        if (snapshot.hasData) {
+          List<Genre> snapshotList = snapshot.data ?? [];
+          movieGenres = repository.getGenresFromId(genresId, snapshotList,);
```
Comment (matiasilveiro-gl) [inline lib/widgets/genres.dart]: Autoformat

File: lib/widgets/genres.dart [hunk 11]
```diff
@@ -1,53 +1,89 @@
...
+          return const CircularProgressIndicator();
+        }
+      },
     );
   }
 }
+
+/*
+
+ */
```
Comment (matiasilveiro-gl) [inline lib/widgets/genres.dart]: Remove this as it's not needed

File: lib/widgets/home_page.dart [hunk 12]
```diff
@@ -1,117 +1,28 @@
...
-                _counter.toString(),
-                style: const TextStyle(
-                  color: Colors.black,
-                ),
-              ),
-            ),
-          ],
-        ),
-      ),
+      body: const SafeArea(child: MovieList()),
```
Comment (matiasilveiro-gl) [inline lib/widgets/home_page.dart]: Trailing coma

File: lib/widgets/genres.dart (hunk 9 above)
Comment (AlejoArea) [reply lib/widgets/genres.dart:26]: I understand the stateful approach, but I will use stateless. Next time I use FutureBuilder I will check whats the best option depending on the application

File: lib/screens/movie_details.dart (hunk 8 above)
Comment (AlejoArea) [reply lib/screens/movie_details.dart]: done

File: lib/repositories/i_repository.dart (hunk 7 above)
Comment (AlejoArea) [reply lib/repositories/i_repository.dart]: Done

File: lib/widgets/genres.dart (hunk 10 above)
Comment (AlejoArea) [reply lib/widgets/genres.dart]: done

File: lib/widgets/genres.dart (hunk 11 above)
Comment (AlejoArea) [reply lib/widgets/genres.dart]: done

File: lib/widgets/home_page.dart (hunk 12 above)
Comment (AlejoArea) [reply lib/widgets/home_page.dart]: done

File: lib/models/movie.dart (hunk 6 above)
Comment (AlejoArea) [reply lib/models/movie.dart]: done

**Review by LeonardoMolleker — APPROVED** [review]

File: lib/widgets/genres.dart (hunk 9 above)
Comment (AlejoArea) [reply lib/widgets/genres.dart]: changed to stateful. Done

File: lib/widgets/genres.dart [hunk 13]
```diff
@@ -1,53 +1,98 @@
...
+  State<Genres> createState() => _GenresState();
+}
 
+class _GenresState extends State<Genres> {
+  late Future<List<Genre>> _allGenres;
+  late List<Genre> movieGenres;
+
+  @override
+  void initState(){
+    _allGenres=widget.repository.getData();
```
Comment (matiasilveiro-gl) [inline lib/widgets/genres.dart]: Spacing

**Review by matiasilveiro-gl — APPROVED** [review]
LGTM!

**Review by MarcosCGlobant — APPROVED** [review]

File: lib/models/movie.dart [hunk 14]
```diff
@@ -1,36 +1,78 @@
...
+      id: 1,
+      adult: false,
       releaseDate: '2023-04-05',
       overview: 'While working underground to fix a water main, '
           'Brooklyn plumbers—and brothers—Mario and Luigi are transported down '
           'a mysterious pipe and wander into a magical new world. But when the '
           'brothers are separated, Mario embarks on an epic quest to find Luigi.',
       voteAverage: 7.8,
-      genres: ['Animation', 'Family', 'Adventure', 'Fantasy', 'Comedy'],
+      genres: [],
```
Comment (FranciscoGaraventa) [inline lib/models/movie.dart]: Why the mock movie have an empty list, here you need to have the IDs from the names that you are removing, right?

Comment (AlejoArea) [reply lib/models/movie.dart]: True I didn't see that because the mock movie isn't used. I will change that

**Review by FranciscoGaraventa — APPROVED** [review]
LG

**Pull request merged**